onnxruntime/core/session/inference_session.cc

```diff
@@ -618,12 +571,9 @@ Status InferenceSession::Run(const RunOptions& run_options,
   Status retval = Status::OK();
 
   try {
-    {
-      std::lock_guard<onnxruntime::OrtMutex> l(session_mutex_);
-      if (!is_inited_) {
-        LOGS(*session_logger_, ERROR) << "Session was not initialized";
-        retval = Status(common::ONNXRUNTIME, common::FAIL, "Session not initialized.");
-      }
+    if (!is_inited_) {
```

Comment:
is_inited_ should be declared std::atomic<bool> if you want to remove the lock here. In fact the C++ standard has a dedicated type for it called atomic_bool. https://en.cppreference.com/w/cpp/atomic/atomic

Reply:
it's only needed when writing access to is_inited_, I think.

Reply:
normally, callers should, 1. create/initialize session. 2) multi-threading call run if he/she's willing to. In that case, it's not needed to have a lock here. in worst case, callers call initialize in separate thread, no lock here will only fail one run call? btw, the lock is there when editing is_inited_, I believe.

Reply:
The safest is to make is_init_ atomic. But I am inclined to believe we are ok here. In Ke's case of calling init in a thread parallel to run call, it is a bug regardless of whatever locks you use.